libcfnet/protocol.c

```diff
@@ -88,3 +89,123 @@ Seq *ProtocolOpenDir(AgentConnection *conn, const char *path)
+
+    unlink(local_path);
+    FILE *file_ptr = safe_fopen_create_perms(local_path, "wx", perms);
+    if (file_ptr == NULL)
+    {
+        Log(LOG_LEVEL_ERR, "%s: Failed to open file %s (fopen: %s)",
+            __func__, local_path, GetErrorStr());
+        return false;
+    }
+
+    char buf[CF_MSGSIZE] = {0};
+    int to_send = snprintf(buf, CF_MSGSIZE, "GET %d %s",
+                           CF_MSGSIZE, remote_path);
+
+    int ret = SendTransaction(conn->conn_info, buf, to_send, CF_DONE);
+    if (ret == -1)
+    {
+        Log(LOG_LEVEL_ERR,
+            "%s: Failed to send request for file %s on host %s",
+            __func__, remote_path, conn->this_server);
+        unlink(local_path);
+        fclose(file_ptr);
+        return false;
+    }
+
+    char cfchangedstr[sizeof(CF_CHANGEDSTR1 CF_CHANGEDSTR2)];
+    snprintf(cfchangedstr, sizeof(cfchangedstr), "%s%s",
+             CF_CHANGEDSTR1, CF_CHANGEDSTR2);
+
+    bool res = true;
+    uint32_t to_recv = 0;
+    while (to_recv < file_size)
+    {
+        int len = TLSRecv(conn->conn_info->ssl, buf, CF_MSGSIZE);
+        if (len == -1)
+        {
+            Log(LOG_LEVEL_ERR,
+                "%s: Failed to retrieve file %s:%s",
+                __func__, conn->this_server, remote_path);
+            res = false;
+            break;
+        }
+        else if (len > CF_MSGSIZE)
+        {
+            Log(LOG_LEVEL_ERR,
+                "%s: Incorrect length of incoming packet "
+                "while retrieving %s:%s, %d > %d",
+                __func__, conn->this_server, remote_path, len, CF_MSGSIZE);
+            res = false;
+            break;
+        }
+
+        if (BadProtoReply(buf))
+        {
+            Log(LOG_LEVEL_ERR,
+                "%s: Error from server while retrieving file %s:%s: %s",
+                __func__, conn->this_server, remote_path, buf);
+            res = false;
+            break;
+        }
+
+        if (strncmp(buf, cfchangedstr, sizeof(cfchangedstr) - 1) == 0)
+        {
+            Log(LOG_LEVEL_ERR,
+                "%s: Remote file %s:%s changed during file transfer",
+                __func__, conn->this_server, remote_path);
+            res = false;
+            break;
+        }
+
+        ret = fwrite(buf, sizeof(char), len, file_ptr);
+        if (ret < 0)
+        {
+            Log(LOG_LEVEL_ERR,
+                "%s: Failed to write during rertieval of file %s:%s",
+                __func__, conn->this_server, remote_path);
+            res = false;
+            break;
+        }
+
+        to_recv += len;
+    }
+
+    if (!res)
+    {
+        unlink(local_path);
+    }
+
+    fclose(file_ptr);
+    return res;
+}
+
+bool ProtocolStatGet(AgentConnection *conn, const char *remote_path,
+                     const char *local_path, int perms)
+{
+    assert(conn != NULL);
+    assert(remote_path != NULL);
+
+    struct stat sb;
+    int ret = cf_remote_stat(conn, false, remote_path, &sb, "file");
```

Comment:
This will be replaced by `ProtocolStat()` right?

Reply:
Yep.